### DISCUSSION TITLE
tutorials/inputs/mouse_and_input_coordinates

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#25] [User-contributed notes]
# tutorials/inputs/mouse_and_input_coordinates

About: The reason for this small tutorial is to clear up many common mistakes about input coordinates, obtaining mouse position and screen resolution, etc. Hardware display coordinates: Using hardw...

https://docs.godotengine.org/en/latest/tutorials/inputs/mouse_and_input_coordinates.html

**Comment by balloonpopper** [comment]
A working example that only fires once when you click a button, and shows you how to check the states of the buttons and mouse wheel : 

```gdscript
func _input(event):
	if event is InputEventMouseButton and event.pressed:
		if event.button_index == MOUSE_BUTTON_LEFT :
			print("Left mouse button")
		if event.button_index == MOUSE_BUTTON_WHEEL_UP :
			print("Scroll wheel up")
		if event.button_index == MOUSE_BUTTON_WHEEL_DOWN:
			print("Scroll wheel down")
```

**Reply by sash-rc** [reply]
Your last two `if`s should be `elif` to discard redundant checks

**Reply by Taywee** [reply]
Or better, a `match`:

```gdscript
func _input(event):
	if event is InputEventMouseButton and event.pressed:
		match event.button_index:
			MOUSE_BUTTON_LEFT:
				print("Left mouse button")
			MOUSE_BUTTON_WHEEL_UP:
				print("Scroll wheel up")
			MOUSE_BUTTON_WHEEL_DOWN:
				print("Scroll wheel down")
```

**Reply by troynall** [reply]
I couldn't get this code to run without putting a colon here " match event.button_index:"
I don't mean to be trivial. Just don't want a new person to get stuck and frustrated.

**Comment by ChronoByteCosmonaut** [comment]
Does someone have a working example of mouse position for Multiplayer cause I can’t get it working! I have tried using get_viewport().get_mouse_position() but it’s way off!! 😅

**Reply by Calinou** [reply]
Are you referring to split screen, or the use of multiple mice at the same time? Godot does not support multiple mice with independent mouse cursors (support for this in operating systems is spotty in general).

If you're referring to split screen, you will need to offset the position returned by `get_mouse_position()` by the viewport's position (and perhaps its scale, if it's scaled).

**Reply by ChronoByteCosmonaut** [reply]
The split screen (server/client ) comm. I found a solution that goes like  : 

" var screen_mouse_position = get_viewport().get_mouse_position()  # Get the mouse position on the screen
	# Convert it to world coordinates	
	mouse_position = (get_viewport().get_screen_transform() * get_viewport().get_canvas_transform()).affine_inverse() * screen_mouse_position " . Now I am struggling with the shooting/killing and respawning players in the game.

**Comment by riki137** [comment]
In case you're trying to convert mouse position to position in your 2D world with a camera, this might be useful:
```gdscript
func _input_event(viewport: Viewport, event: InputEvent, shape_idx: int) -> void:
    if event is InputEventMouseButton and event.pressed:
        var camera = viewport.get_camera_2d()
        click.emit((event.position / camera.zoom) + camera.offset)
```

EDIT: this proved to work the same??
```gdscript
func _input(event: InputEvent) -> void:
     if event is InputEventMouseButton and event.pressed:
        var camera = get_viewport().get_camera_2d()
        click.emit(camera.get_global_mouse_position())
```

**Reply by idbrii** [reply]
`get_camera_2d` is so helpful! Makes it very easy to make my gamepad direction selecting code to also work for mouse!

```gdscript
func _input(event: InputEvent):
	if event is InputEventMouseMotion:
		var camera = get_viewport().get_camera_2d()
		var direction: Vector2 = camera.get_global_mouse_position() - global_position
		# Normalize to treat any mouse movement makes as valid positioning.
		handle_input(direction.normalized())


func _process(_dt: float):
	var motion: Vector2 = Input.get_vector("move_left", "move_right", "move_up", "move_down")
	handle_input(motion)


func handle_input(motion: Vector2):
	if not motion.is_zero_approx():
        # My stored aim direction.
		requested_motion = motion

	destination_pointer.rotation = requested_motion.angle()

	if Input.is_action_just_pressed("move_confirm"):
        # Do tween to move in requested_motion direction.
        ...
```

**Comment by GuyCooke** [comment]
There is a function on ```CanvasItem``` called ```make_input_local``` that takes an event passed to ```_input``` which is in screen coordinates and gives you a new one in local coordinates for any ```CanvasItem``` you please. Very handy and woefully hard to find when you're looking into ```InputEvent```s.

I actually think there should be a function, ```InputEvent.make_input_local(canvas_item: CanvasItem)``` that just does

```return canvas_item.make_input_local(self)```

just so folks would find it when they're looking for options with ```InputEvent``` instances.

**Comment by Th9Fan2** [comment]
additional if you what the mouse position to max out at the size of the screen i made this :

```gdscript
viewport = get_viewport().size
mouse = get_viewport().get_mouse_position()
if mouse[0] > viewport[0]:
	mouse[0] = viewport[0]
if mouse[1] > viewport[1]:
	mouse[1] = viewport[1]
if mouse[0] < 0:
	mouse[0] = 0
if mouse[1] < 0:
	mouse[1] = 0
print(viewport)
print(mouse)
```

**Comment by brainy19** [comment]
I'm trying to code a "demolish" feature in my game, but nothing is happening. This is what I have so far:

```gdscript
func _on_demolish_pressed(event):
	if InputEventMouseButton and event.pressed:
		if event.button_index == MOUSE_BUTTON_LEFT:
			ground.erase_cell(event.position())
```

"ground" is the tile map layer that has the buildings being built on.

**Reply by Calinou** [reply]
You need to map the mouse coordinates (which are in pixels) to tile coordinates using [`TileMapLayer.local_to_map()`](https://docs.godotengine.org/en/latest/classes/class_tilemaplayer.html#class-tilemaplayer-method-local-to-map). Read that method's description for details, especially when the input coordinates are global.

**Comment by troynall** [comment]
1. *Incorrect Event Check*
   - if InputEventMouseButton is not checking if event is actually a mouse button event.
     Instead, use if event is InputEventMouseButton:.
   
2. *Position Handling*
   - event.position() does not exist.
     The correct way to get the position of a mouse click is event.position
   
3. *Erasing the Tile in the TileMap*
   - erase_cell() requires coordinates in tile space, not pixel space.
      You need to convert event.position (which is in global screen
      coordinates) into tile coordinates using local_to_map().

Fixed Code:
func _on_demolish_pressed(event):
    if event is InputEventMouseButton and event.pressed:
        if event.button_index == MOUSE_BUTTON_LEFT:
            var tile_pos = ground.local_to_map(event.position)
            ground.erase_cell(0, tile_pos)  # Assuming layer 0

Explanation of Fixes:

   -event is InputEventMouseButton ensures we're handling a mouse click event.
   -event.position gives the correct mouse position.
   -local_to_map(event.position) converts the mouse position to tile coordinates.
   -erase_cell(0, tile_pos) removes the tile at the given position on *layer 0*.

If it's still not working, make sure:
-ground is properly referenced and set in your script.
-The tile map actually has tiles placed before erasing.
-_on_demolish_pressed() is correctly connected to the input event signal.

Additional Debugging:

If it still doesn’t work *Ensure the function is connected*
- If you're using a *Button* node, _on_demolish_pressed() won't receive mouse events directly.
-Instead, use _input(event) inside a script attached to the TileMap node:

**Reply by sarisman84** [reply]
Please format your comment. It is borderline unreadable.
